e2e: ceph-csi-operator deployment support

iPraveenParihar opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

This PR supports running e2e tests against a ceph-csi deployment via [ceph-csi-operator](https://github.com/ceph/ceph-csi-operator).

Includes - 
- scripts/deploy-ceph-csi-operator.sh to deploy ceph-csi via operator.
- e2e/operator.go includes util methods for ceph-csi-operator.
- changes to handle delete/update of ceph-csi pods.

## Is there anything that requires special attention ##

Do you have any questions?

Is the change backward compatible?

Are there concerns around backward compatibility?

Provide any external context for the change, if any.

For example:

* Kubernetes links that explain why the change is required
* CSI spec related changes/catch-up that necessitates this patch
* golang related practices that necessitates this change

## Related issues ##

Mention any github issues relevant to this PR. Adding below line
will help to auto close the issue once the PR is merged.

Depends-on: #4931
Fixes: #4856 

## Future concerns ##

List items that are not part of the PR and do not impact it's
functionality, but are work items that can be taken up subsequently.

**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

iPraveenParihar committed:
ci: script to deploy ceph-csi via ceph-csi-operator

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
e2e: handle ceph-csi-operator deployment changes

This commits adds e2e/operator.go containing utility
methods specific to the operator.

Signed-off-by: Praveen M <m.praveen@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

nixpanic commented:
> /test ci/centos/mini-e2e-operator/k8s-1.31

This failed to pull the `quay.io/cephcsi/ceph-csi-operator:` (without tag?) from the CI mirror.  You will need to add the container-image to [`images.txt`](https://github.com/ceph/ceph-csi/blob/ci/centos/mirror/images.txt) and wait until tomorrow to try again.

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
/retest ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
/retest ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar committed:
ci: use ceph-csi-operator latest tag

Use the main branch instead of the release till the following fixes
are available in next release (> v0.2.0)
- https://github.com/ceph/ceph-csi-operator/pull/206
- https://github.com/ceph/ceph-csi-operator/pull/207

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
operator CI failed - [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e-operator_k8s-1.31/runs/15/nodes/102/log/?start=0) 

- CephFS failed at fuse recovery test as the mount-info volumeMount is not present (_raised a PR on ceph-csi-operator_)
- NFS: failed  `subvolume count 0 not matching expected count 1` (🤔 it passed on my machine.. probably updated csi-config-map didn't take in effect when test ran.)

iPraveenParihar commented:
/retest ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
failed ci [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e-operator_k8s-1.31/runs/16/nodes/102/log/?start=0)

```
  [38;5;9m[FAILED] failed to test volumeGroupSnapshot: failed to create volume group snapshot: failed to get VolumeGroupSnapshot cephfs-3893-vgs: failed to get volumesnapshot: client rate limiter Wait returned an error: context deadline exceeded[0m
  ```

iPraveenParihar commented:
/retest ci/centos/mini-e2e-operator/k8s-1.31

Madhu-1 commented:
> failed to test volumeGroupSnapsho

@iPraveenParihar it looks like the volumegroupsnapshot policy is not enabled in the Driver/OperatorConfig CR https://github.com/ceph/ceph-csi-operator/blob/6e1c39c697cb397119920bff6551d12ceeb1a514/api/v1alpha1/driver_types.go#L308

iPraveenParihar commented:
> @iPraveenParihar it looks like the volumegroupsnapshot policy is not enabled in the Driver/OperatorConfig CR https://github.com/ceph/ceph-csi-operator/blob/6e1c39c697cb397119920bff6551d12ceeb1a514/api/v1alpha1/driver_types.go#L308

Yes, thanks I noticed it 😀 . Testing on local machine

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
```
I0227 09:35:32.750840       1 utils.go:266] ID: 22 Req-ID: pvc-844f6615-fdee-4b13-b063-2e563b69f91c GRPC call: /csi.v1.Controller/CreateVolume
I0227 09:35:32.751600       1 utils.go:267] ID: 22 Req-ID: pvc-844f6615-fdee-4b13-b063-2e563b69f91c GRPC request: {"capacity_range":{"required_bytes":1073741824},"name":"pvc-844f6615-fdee-4b13-b063-2e563b69f91c","parameters":{"clusterID":"e809629f-2772-4bac-8592-4575ebe7f2c7","fsName":"myfs","nfsCluster":"my-nfs","server":"rook-ceph-nfs-my-nfs-a.rook-ceph.svc.cluster.local","volumeNamePrefix":"nfs-export-"},"secrets":"***stripped***","volume_capabilities":[{"access_mode":{"mode":"SINGLE_NODE_SINGLE_WRITER"},"mount":{"fs_type":"ext4"}}]}
W0227 09:35:32.751628       1 credentials.go:119] adminID and adminKey are deprecated, please use userID and userKey instead
I0227 09:35:32.756124       1 omap.go:89] ID: 22 Req-ID: pvc-844f6615-fdee-4b13-b063-2e563b69f91c got omap values: (pool="myfs-metadata", namespace="csi", name="csi.volumes.default"): map[]
W0227 09:35:32.756146       1 credentials.go:119] adminID and adminKey are deprecated, please use userID and userKey instead
I0227 09:35:32.763523       1 omap.go:159] ID: 22 Req-ID: pvc-844f6615-fdee-4b13-b063-2e563b69f91c set omap keys (pool="myfs-metadata", namespace="csi", name="csi.volumes.default"): map[csi.volume.pvc-844f6615-fdee-4b13-b063-2e563b69f91c:de017c04-4b39-4197-835a-1a0ca60974a6])
I0227 09:35:32.767254       1 omap.go:159] ID: 22 Req-ID: pvc-844f6615-fdee-4b13-b063-2e563b69f91c set omap keys (pool="myfs-metadata", namespace="csi", name="csi.volume.de017c04-4b39-4197-835a-1a0ca60974a6"): map[csi.imagename:nfs-export-de017c04-4b39-4197-835a-1a0ca60974a6 csi.volname:pvc-844f6615-fdee-4b13-b063-2e563b69f91c])
I0227 09:35:32.767276       1 fsjournal.go:318] ID: 22 Req-ID: pvc-844f6615-fdee-4b13-b063-2e563b69f91c Generated Volume ID (0001-0024-e809629f-2772-4bac-8592-4575ebe7f2c7-0000000000000001-de017c04-4b39-4197-835a-1a0ca60974a6) and subvolume name (nfs-export-de017c04-4b39-4197-835a-1a0ca60974a6) for request name (pvc-844f6615-fdee-4b13-b063-2e563b69f91c)
I0227 09:35:32.822304       1 controllerserver.go:475] ID: 22 Req-ID: pvc-844f6615-fdee-4b13-b063-2e563b69f91c cephfs: successfully created backing volume named nfs-export-de017c04-4b39-4197-835a-1a0ca60974a6 for request name pvc-844f6615-fdee-4b13-b063-2e563b69f91c
I0227 09:35:32.822355       1 controllerserver.go:89] ID: 22 Req-ID: pvc-844f6615-fdee-4b13-b063-2e563b69f91c CephFS volume created: 0001-0024-e809629f-2772-4bac-8592-4575ebe7f2c7-0000000000000001-de017c04-4b39-4197-835a-1a0ca60974a6
W0227 09:35:32.822364       1 credentials.go:119] adminID and adminKey are deprecated, please use userID and userKey instead
I0227 09:35:32.827879       1 omap.go:159] ID: 22 Req-ID: pvc-844f6615-fdee-4b13-b063-2e563b69f91c set omap keys (pool="myfs-metadata", namespace="csi", name="csi.volume.de017c04-4b39-4197-835a-1a0ca60974a6"): map[csi.nfs.cluster:my-nfs])
I0227 09:35:32.946820       1 controllerserver.go:116] ID: 22 Req-ID: pvc-844f6615-fdee-4b13-b063-2e563b69f91c published NFS-export: 0001-0024-e809629f-2772-4bac-8592-4575ebe7f2c7-0000000000000001-de017c04-4b39-4197-835a-1a0ca60974a6
I0227 09:35:32.946924       1 utils.go:273] ID: 22 Req-ID: pvc-844f6615-fdee-4b13-b063-2e563b69f91c GRPC response: {"volume":{"capacity_bytes":1073741824,"volume_context":{"backingSnapshot":"false","clusterID":"e809629f-2772-4bac-8592-4575ebe7f2c7","fsName":"myfs","nfsCluster":"my-nfs","server":"rook-ceph-nfs-my-nfs-a.rook-ceph.svc.cluster.local","share":"/0001-0024-e809629f-2772-4bac-8592-4575ebe7f2c7-0000000000000001-de017c04-4b39-4197-835a-1a0ca60974a6","subvolumeName":"nfs-export-de017c04-4b39-4197-835a-1a0ca60974a6","subvolumePath":"/volumes/e2e/nfs-export-de017c04-4b39-4197-835a-1a0ca60974a6/c018210b-da7d-4458-a0ba-91eb9fc94b9a","volumeNamePrefix":"nfs-export-"},"volume_id":"0001-0024-e809629f-2772-4bac-8592-4575ebe7f2c7-0000000000000001-de017c04-4b39-4197-835a-1a0ca60974a6"}}
```

NFS test failed - 

For PVC - ` pvc-844f6615-fdee-4b13-b063-2e563b69f91c` is created in `e2e` subvolumegroup and expected to be in `csi`. 

```
I0227 09:35:32.946924       1 utils.go:273] ID: 22 Req-ID: pvc-844f6615-fdee-4b13-b063-2e563b69f91c GRPC response: {"volume":{"capacity_bytes":1073741824,"volume_context":{"backingSnapshot":"false","clusterID":"e809629f-2772-4bac-8592-4575ebe7f2c7","fsName":"myfs","nfsCluster":"my-nfs","server":"rook-ceph-nfs-my-nfs-a.rook-ceph.svc.cluster.local","share":"/0001-0024-e809629f-2772-4bac-8592-4575ebe7f2c7-0000000000000001-de017c04-4b39-4197-835a-1a0ca60974a6","subvolumeName":"nfs-export-de017c04-4b39-4197-835a-1a0ca60974a6","subvolumePath":"/volumes/e2e/nfs-export-de017c04-4b39-4197-835a-1a0ca60974a6/c018210b-da7d-4458-a0ba-91eb9fc94b9a","volumeNamePrefix":"nfs-export-"},"volume_id":"0001-0024-e809629f-2772-4bac-8592-4575ebe7f2c7-0000000000000001-de017c04-4b39-4197-835a-1a0ca60974a6"}}
```

For NFS BeforeEach call, we create a ConfigMap with `csi` subvolumegroup. Probably, there was a significant delay in updated `ceph-csi-config` ConfigMap to be available in csi pods. 

Similar thing happened for CephFS test `verify rados objects are within a namespace `

```
0227 09:33:27.905137       1 omap.go:159] ID: 257 Req-ID: pvc-11fc74aa-842b-47ab-b73f-95d461cce56a set omap keys (pool="myfs-metadata", namespace="csi", name="csi.volumes.default"): map[csi.volume.pvc-11fc74aa-842b-47ab-b73f-95d461cce56a:64a5c26a-52de-48ed-8511-7493db12421d])
I0227 09:33:27.909233       1 omap.go:159] ID: 257 Req-ID: pvc-11fc74aa-842b-47ab-b73f-95d461cce56a set omap keys (pool="myfs-metadata", namespace="csi", name="csi.volume.64a5c26a-52de-48ed-8511-7493db12421d"): map[csi.imagename:csi-vol-64a5c26a-52de-48ed-8511-7493db12421d csi.volname:pvc-11fc74aa-842b-47ab-b73f-95d461cce56a csi.volume.owner:cephfs-6326])
I0227 09:33:27.909253       1 fsjournal.go:318
```

```
  [38;5;9m[FAILED] failed to validate omap count for rados ls --pool=myfs-metadata --namespace=cephfs-ns | grep -v default | grep -v csi.volume.group. | grep -c ^csi.volume.: expected omap object count 2, got 0[0m
  ```

I'll try to scale down the csi controller pods before updating/creating ConfigMap and then scale up.

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.32

iPraveenParihar commented:
operator deployment CI passed [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e-operator_k8s-1.32/detail/mini-e2e-operator_k8s-1.32/1/pipeline), 

helm CI test failed due to label selector not provided. [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/job/mini-e2e-helm_k8s-1.31/371/display/redirect)

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.32

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.30

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
Helm E2E inconsistent failure observed -
 
```
  [38;5;9m[FAILED] timeout waiting for deployment csi-rbdplugin-provisioner: error waiting for deployment "csi-rbdplugin-provisioner" status to match desired state: client rate limiter Wait returned an error: rate: Wait(n=1) would exceed context deadline[0m
  [38;5;9mIn [1m[BeforeEach][0m[38;5;9m at: [1m/go/src/github.com/ceph/ceph-csi/e2e/rbd.go:364[0m [38;5;243m@ 02/28/25 14:51:41.74[0m
```

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.30

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
passed E2E logs - 
- [ci/centos/mini-e2e/k8s-1.30](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/job/mini-e2e_k8s-1.30/533/display/redirect)
- [ci/centos/mini-e2e-operator/k8s-1.30](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/job/mini-e2e-operator_k8s-1.30/2/display/redirect)
- [ci/centos/mini-e2e-helm/k8s-1.30](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/job/mini-e2e-helm_k8s-1.30/523/display/redirect)
- [ci/centos/mini-e2e-helm/k8s-1.31](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/job/mini-e2e-helm_k8s-1.31/388/display/redirect)

iPraveenParihar commented:
@Rakshith-R, could you please take some time and provide review on this? Thanks!

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.30

Rakshith-R reviewed (dismissed):
Thanks

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -266,11 +285,12 @@ var _ = Describe(cephfsType, func() {
-					framework.Failf("failed to delete namespace %s: %v", cephCSINamespace, err)
-				}
+		}
+		// No need to create the namespace if ceph-csi is deployed via helm or operator.
```
Please update the comment here

Madhu-1 commented on e2e/configmap.go:
```diff
@@ -93,6 +93,9 @@ func createConfigMap(pluginPath string, c kubernetes.Interface, f *framework.Fra
 	}
 	if apierrs.IsNotFound(err) {
 		_, err = c.CoreV1().ConfigMaps(cephCSINamespace).Create(context.TODO(), &cm, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
```
Is this required.  there is as err returning already at the end of the function

Madhu-1 commented on e2e/e2e_test.go:
```diff
@@ -52,6 +52,7 @@ func init() {
 	flag.StringVar(&fileSystemName, "filesystem", "myfs", "CephFS filesystem to use")
 	flag.StringVar(&clusterID, "clusterid", "", "Ceph cluster ID to use (defaults to `ceph fsid` detection)")
 	flag.StringVar(&nfsDriverName, "nfs-driver", "nfs.csi.ceph.com", "name of the driver for NFS-volumes")
+	flag.BoolVar(&operatorDeployment, "operator-deployment", false, "test running on deployment via operator")
```
```suggestion
flag.BoolVar(&operatorDeployment, "operator-deployment", false, "deploy ceph-csi using the operator")
```

nixpanic replied:
I am missing why this is needed. But maybe that becomes clear in a next commit (could be squash together)?

Madhu-1 commented on scripts/deploy-ceph-csi-operator.sh:
```diff
@@ -0,0 +1,148 @@
+    name: ${IMAGESET_CONFIGMAP_NAME}
+    namespace: ${OPERATOR_NAMESPACE}
+data:
+    "plugin": "quay.io/cephcsi/cephcsi:canary"  # test image
```
get the cephcsi version from the build.env?

Madhu-1 commented on scripts/deploy-ceph-csi-operator.sh:
```diff
@@ -0,0 +1,148 @@
+
+function deploy_operator() {
+    kubectl_retry create -f "${OPERATOR_INSTALL}"
+    create_or_delete_operator_config "create"
```
define a var for create and delete as its used in multiple places

nixpanic replied:
or use `create` and `delete` only once :wink:

```
    generate_operator_config
    generate_driver $RBD_DRIVER_NAME
    generate_driver $CEPHFS_DRIVER_NAME
    generate_driver $NFS_DRIVER_NAME
    kubectl_retry "create" -f "${TEMP_DIR}"
```

Please try to use the same coding standard everywhere, like `"${RBD_DRIVER_NAME}"` with the quotes and such.

nixpanic commented on scripts/deploy-ceph-csi-operator.sh:
```diff
@@ -0,0 +1,143 @@
+
+function create_or_delete_imageset_configmap() {
+    local operation=$1
+    temp_file=$(mktemp "${TEMP_DIR}/imageset-configmap.XXXXXX.yaml")
```
You place all these files in a tempdir, there is no need to create filenames with the XXXXXX pattern there. It might be nice to generate all files, display the contents (useful for debugging) and then `kubectl create -f "${TEMP_DIR}"` everything at once.

This also makes the workflow a little simpler; the `kubectl` command does need to be run at the end of every function, so no `create_or_delete` prefix and `operation` needed.

nixpanic commented on e2e/cephfs.go:
```diff
@@ -175,13 +183,19 @@ var _ = Describe(cephfsType, func() {
+			cephFSDeamonSetName = operatorCephFSDaemonsetName
+		}
+		// No need to create the namespace if ceph-csi is deployed via helm or operator.
+		if cephCSINamespace != defaultNs && (!helmTest && !operatorDeployment) {
```
personally I think `!(helmTest || operatorDeployment)` reads a little easier

nixpanic commented on e2e/nfs.go:
```diff
@@ -242,14 +262,21 @@ var _ = Describe("nfs", func() {
-				if err != nil {
-					framework.Failf("failed to create namespace %s: %v", cephCSINamespace, err)
-				}
+		if operatorDeployment {
```
I'm starting to get annoyed by this `operatorDeployment` check, it seems to come up a lot. This suggests to me that it should get improved to increase maintainability.

Consider something like this:

```go
// in e2e/e2e_test.go maybe

type DeploymentMethod struct {
    nfsDeploymentName string
    nfsDeamonSetName string
    // and all other "if operatorDeployment" things
}

var deploymentMethod *DeploymentMethod = nil

if operatorDeployment {
    operatorDeployment = &DeploymentMethod{
        nfsDeploymentName: operatorNFSDeploymentName,
        //...
    }
} else {
    operatorDeployment = &DeploymentMethod{
        nfsDeploymentName: nfsDeploymentName,
        //...
    }
}
```

With that, you can just use `deploymentMethod.nfsDeploymentName` everywhere, and no `if operatorDeployment` needed, except for the initial creation of the struct.

You could even make it more easier to use by adding functions like `operatorDeployment.HasNFS()` which returns the (private?) `deployNFS` bool. Or give it a `operatorDeployment.DeployNFS()` function that does nothing in case `deployNFS` is not set.

nixpanic commented on e2e/operator.go:
```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2024 The Ceph-CSI Authors.
```
I know you started this last year, but you may update it to 2025 if you like.

nixpanic commented on e2e/rbd.go:
```diff
@@ -290,7 +297,24 @@ var _ = Describe("RBD", func() {
+			rbdDeploymentName = operatorRBDDeploymentName
+			rbdDaemonsetName = operatorRBDDaemonsetName
+
+			err := setDomainLabels([]string{nodeRegionLabel, nodeZoneLabel})
```
Steps like this should probably be part of a `DeploymentMethod` struct or interface. All the `if operatorDeployment` and `if !helmTest` make it very confusing to follow this code while reading it. Possibly things like this can be based on the `ResourceDeployer` interface, and a list of those structs can be executed depending on the `DeploymentMethod`.

nixpanic commented on scripts/deploy-ceph-csi-operator.sh:
```diff
@@ -7,7 +7,12 @@ SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+# - https://github.com/ceph/ceph-csi-operator/pull/207
+# Till then use the main branch instead of the release.
+# OPERATOR_VERSION=${OPERATOR_VERSION:-"main"}
+OPERATOR_VERSION="main"
```
can you not just set this (with the comment) in `build.env`? It is cleaner to update the version there than in a script.

iPraveenParihar replied:
This will require us to add main tag in ci/centos branch

```
+ ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@n27-39-115.pool.ci.centos.org 'podman pull --authfile=~/.podman-auth.json registry-ceph-csi.apps.ocp.cloud.ci.centos.org/cephcsi/ceph-csi-operator:main && podman tag registry-ceph-csi.apps.ocp.cloud.ci.centos.org/cephcsi/ceph-csi-operator:main cephcsi/ceph-csi-operator:main quay.io/cephcsi/ceph-csi-operator:main'

Trying to pull registry-ceph-csi.apps.ocp.cloud.ci.centos.org/cephcsi/ceph-csi-operator:main...

Error: initializing source docker://registry-ceph-csi.apps.ocp.cloud.ci.centos.org/cephcsi/ceph-csi-operator:main: reading manifest main in registry-ceph-csi.apps.ocp.cloud.ci.centos.org/cephcsi/ceph-csi-operator: manifest unknown

script returned exit code 125
```

nixpanic reviewed (commented):
Happy to see this works, well done!

Unfortunately all the extra if-statements in the e2e suite make it even more hard to maintain. It would be best to clean that up, introduce a mechanism that can be configured in a single place, and call all possibly needed functions later (such functions could be a no-op for certain deployments).

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.30

iPraveenParihar commented:
@nixpanic, I have addressed your reviews. PTAL
Thanks!

nixpanic commented on scripts/deploy-ceph-csi-operator.sh:
```diff
@@ -0,0 +1,142 @@
+# shellcheck disable=SC1091
+source "${SCRIPT_DIR}/../build.env"
+
+OPERATOR_VERSION=${OPERATOR_VERSION:-"main"}
```
should this not be `CEPH_CSI_OPERATOR_VERSION`?

iPraveenParihar replied:
Thanks! will take care of it,

nixpanic commented on scripts/deploy-ceph-csi-operator.sh:
```diff
@@ -0,0 +1,142 @@
+    "snapshotter": "${K8S_IMAGE_REPO}/csi-snapshotter:${CSI_SNAPSHOTTER_VERSION}"
+    "provisioner": "${K8S_IMAGE_REPO}/csi-provisioner:${CSI_PROVISIONER_VERSION}"
+    "registrar": "${K8S_IMAGE_REPO}/csi-node-driver-registrar:${CSI_NODE_DRIVER_REGISTRAR_VERSION}"
+    "resizer": "${K8S_IMAGE_REPO}/csi-resizer:${CSI_RESIZER_VERSION}"
```
what about the csi-addons sidecar?

iPraveenParihar replied:
not needed as we don't have csi-addons related tests.

nixpanic commented on e2e/cephfs.go:
```diff
@@ -46,8 +46,41 @@ var (
+	return nil
+}
+
+func (c *CephFSDeployment) getDeploymentName() string {
```
Make these functions part of the `DriverInfo` struct, so that there is no need to have them for each deployment type.

iPraveenParihar replied:
This won't work because `DriverInfo` is embedded in `RBDDeployment`, `CephFSDeployment`, and `NFSDeployment`, but their respective interfaces (`RBDDeploymentMethod`, `CephFSDeploymentMethod`, `NFSDeploymentMethod`) do not inherit methods like `getDeploymentName()` and `getDaemonsetName()` from `DriverInfo`.

For best, I can do is as below
```go
...
type DeploymentMethod interface {
	getDeploymentName() string
	getDaemonsetName() string
}

type RBDDeploymentMethod interface {
	DeploymentMethod
	setClusterName(clusterName string) error
	setDomainLabels(labels []string) error
	setEnableMetadata(value bool) error
	getPodSelector() string
}
...
```
Do you have any alternative approaches in mind? Or did I got it wrong?

nixpanic replied:
Right, interfaces do not inherit the implementation of the functions.

You can do it like
```go
type deploymentMethod struct {
	daemonsetName string
}

func (dm *deploymentMethod) getDaemonsetName() string {
	return dm.daemonsetName
}

type CephFSDeployment struct {
	DriverInfo
	*deploymentMethod
}

func NewCephFSDeployment(...) DeploymentMethod {
	return &CephFSDeployment{
		DriverInfo: DriverInfo{
			...,
		},
		deploymentMethod: &deploymentMethod{
			daemonsetName: ...,
		},
	}
}
```

Or, maybe you want to add functions lile `setDaemonsetName()` to the `DeploymentMethod` interface and `deploymentMethod` struct.

nixpanic replied:
oh, wait, maybe my suggestion does not fit well...

Your example code should work, if the `DriverInfo` struct implements the `DeploymentMethod` function. You can add a type assertion for that:

```go
var _ DeploymentMethod = DriverInfo{}
```

nixpanic reviewed (commented):
needs a few minor adjustments, looks good to me otherwise.

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.30

iPraveenParihar commented:
> /test ci/centos/mini-e2e-operator/k8s-1.30

passed CI [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/job/mini-e2e-operator_k8s-1.30/11/display/redirect)

nixpanic commented on e2e/utils.go:
```diff
@@ -80,33 +77,73 @@ const (
+	setClusterName(clusterName string) error
+	setDomainLabels(labels []string) error
+	setEnableMetadata(value bool) error
+	getPodSelector() string
```
`getPodSelector()` and `setClusterName()` can be part of the `DeploymentMethod` interface?

iPraveenParihar replied:
Done

nixpanic reviewed (commented):
whats up with the **ci: hack commit**?

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.30

iPraveenParihar commented:
> whats up with the ci: hack commit?

@nixpanic, the hack commit was to bypass this error for time being - 
```
Error: initializing source docker://registry-ceph-csi.apps.ocp.cloud.ci.centos.org/cephcsi/ceph-csi-operator:latest: reading manifest latest in registry-ceph-csi.apps.ocp.cloud.ci.centos.org/cephcsi/ceph-csi-operator: manifest unknown

script returned exit code 125
```

let's add the latest tag to `ci/centos/mirror/images.txt` ?

nixpanic reviewed (dismissed):
Tomorrow the `:latest` version should be included in the container-registry that is used in the CentOS CI (updated nightly). Tests are expected to pass with that.

iPraveenParihar commented:
/test ci/centos/mini-e2e-operator/k8s-1.30

iPraveenParihar commented:
> Tomorrow the `:latest` version should be included in the container-registry that is used in the CentOS CI (updated nightly). Tests are expected to pass with that.

[CI](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/job/mini-e2e-operator_k8s-1.30/13/display/redirect) passed ✅

iPraveenParihar commented:
@Rakshith-R @Madhu-1 PTAL.
needs one more review & approval 😀

Rakshith-R commented on scripts/deploy-ceph-csi-operator.sh:
```diff
@@ -0,0 +1,146 @@
+    echo "  $0 cleanup" >&2
+    exit 1
+    ;;
+esac
```
a new line is required here.

iPraveenParihar replied:
Done

Rakshith-R reviewed (dismissed):
Thanks !

iPraveenParihar commented:
@nixpanic, there was a small change can you give approval again?

iPraveenParihar commented:
@Mergifyio rebase

Rakshith-R commented:
@mergifyio queue

iPraveenParihar commented:
/test ci/centos/k8s-e2e-external-storage/1.29

iPraveenParihar commented:
/test ci/centos/k8s-e2e-external-storage/1.30

iPraveenParihar commented:
/test ci/centos/k8s-e2e-external-storage/1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.29

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.29

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
/test ci/centos/upgrade-tests-cephfs

iPraveenParihar commented:
/test ci/centos/upgrade-tests-rbd

iPraveenParihar commented:
Had forgot to make changes in upgrade_rbd.go and upgrade_cephfs.go 😞

iPraveenParihar commented:
/test ci/centos/upgrade-tests-rbd

iPraveenParihar commented:
/test ci/centos/upgrade-tests-cephfs

nixpanic commented on e2e/upgrade-cephfs.go:
```diff
@@ -61,7 +61,27 @@ var _ = Describe("CephFS Upgrade Testing", func() {
 		}
 		c = f.ClientSet
-		if cephCSINamespace != defaultNs {
+		cephFSDeployment = &CephFSDeployment{
```
This is copied into 2 places now, please make a `NewCephFSDeployment()` function, and one for RBD too.

iPraveenParihar commented:
/test ci/centos/upgrade-tests

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.30

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.32

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.31

iPraveenParihar commented:
/test ci/centos/mini-e2e/k8s-1.30

iPraveenParihar commented:
/test ci/centos/k8s-e2e-external-storage/1.32

iPraveenParihar commented:
/test ci/centos/k8s-e2e-external-storage/1.31

iPraveenParihar commented:
/test ci/centos/k8s-e2e-external-storage/1.30

iPraveenParihar commented:
/test ci/centos/upgrade-tests

nixpanic commented:
@mergifyio rebase

nixpanic commented:
/test ci/centos/mini-e2e-helm/k8s-1.32

nixpanic commented:
/test ci/centos/mini-e2e-helm/k8s-1.31

nixpanic commented:
/test ci/centos/mini-e2e-helm/k8s-1.30

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.32

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.31

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.30

nixpanic commented:
/test ci/centos/k8s-e2e-external-storage/1.32

nixpanic commented:
/test ci/centos/k8s-e2e-external-storage/1.31

nixpanic commented:
/test ci/centos/k8s-e2e-external-storage/1.30

nixpanic commented:
/test ci/centos/upgrade-tests

iPraveenParihar commented:
/retest ci/centos/mini-e2e/k8s-1.32
